flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/AbstractAsyncStateStreamOperator.java

```diff
@@ -345,6 +365,56 @@ public void processWatermarkStatus(WatermarkStatus watermarkStatus) throws Excep
                 () -> super.processWatermarkStatus(watermarkStatus));
     }
 
+    @Override
+    protected void processWatermarkStatus(WatermarkStatus watermarkStatus, int index)
+            throws Exception {
+        if (!isAsyncStateProcessingEnabled()) {
+            super.processWatermarkStatus(watermarkStatus, index);
+            return;
+        }
+        asyncExecutionController.processNonRecord(
+                () -> {
+                    boolean wasIdle = combinedWatermark.isIdle();
+                    if (combinedWatermark.updateStatus(index - 1, watermarkStatus.isIdle())) {
```

Comment:
Should it be `index`, instead of `index - 1`?

And can we directly use `() -> super.processWatermarkStatus(...)` here?

Reply:
> Should it be index, instead of index - 1?
Yes, it should be `index` in `AbstractAsyncStateStreamOperator` .

> can we directly use () -> super.processWatermarkStatus(...) here?

No, `super.processWatermarkStatus(...)` will call `processWatermark()`  which is overridded by `AbstractAsyncStateStreamOperator`.